File: solutions/4176-minimum-k-to-reduce-array-within-limit/minimum-k-to-reduce-array-within-limit.cpp
class Solution {
public:
    int nonPositive(const std::vector<int>& a, int k) {
        int result = 0;
        for (int x : a)
            result += (x + k - 1) / k;
        return result;
    }

    int minimumK(std::vector<int>& nums) {
        // Filter <= 0
        std::vector<int> a;
        a.reserve(nums.size());
        for (int x : nums)
            if (x > 0) 
                a.push_back(x);
        if (a.empty()) 
            return 0;

        int l = 1, r = std::max(*max_element(a.begin(), a.end()), (int)a.size());
        while (l < r) {
            int m = l + (r - l) / 2;
            long long mm = (long long)m * (long long)m;
            if (nonPositive(a, m) <= mm)
                r = m;
            else
                l = m + 1;
        }
        return l;
    }
};

File: solutions/4176-minimum-k-to-reduce-array-within-limit/minimum-k-to-reduce-array-within-limit.java
class Solution {
    private long nonPositive(int[] a, int n, int k) {
        long ops = 0;
        for (int i = 0; i < n; i++)
            ops += (a[i] + k - 1L) / k;
        return ops;
    }

    public int minimumK(int[] nums) {
        int n0 = nums.length;

        // Filter <= 0
        int[] a = new int[n0];
        int n = 0, mx = 0;
        for (int x : nums) {
            if (x > 0) {
                a[n++] = x;
                if (x > mx) 
                    mx = x;
            }
        }
        if (n == 0) 
            return 0;

        int l = 1, r = mx > n ? mx : n;

        while (l < r) {
            int m = l + (r - l) / 2;
            long ops = nonPositive(a, n, m);
            long mm = 1L * m * m;
            if (ops <= mm) 
                r = m;
            else 
                l = m + 1;
        }
        return l;
    }
}